Mathlib/NumberTheory/EllipticDivisibilitySequence.lean

```diff
@@ -65,26 +75,506 @@ elliptic, divisibility, sequence
+private def addMulSub (m n : ℤ) : R := W ((m + n).div 2) * W ((m - n).div 2)
+/- Implementation note: we use `Int.div _ 2` instead of `_ / 2` so that `(-m).div 2 = -(m.div 2)`
+and lemmas like `addMulSub_neg_left` hold unconditionally, even though in the case we care about
+(`m` and `n` both even or both odd) both are equal. -/
+
+/-- The four-index elliptic relation, defined in terms of `addMulSub`,
+featuring the three partitions of four indices into two pairs.
+Intended to apply to four integers of the same parity. -/
+private def rel₄ (a b c d : ℤ) : R :=
+  addMulSub W a b * addMulSub W c d
+    - addMulSub W a c * addMulSub W b d
+    + addMulSub W a d * addMulSub W b c
+
+/-- The defining property of elliptic nets in [Stange2011],
+equivalent to a suitable valid (same-parity indices) `rel₄` relation,
+but here only the first three indices enjoy symmetry under permutation,
+while in `rel₄` all four indices can be freely permuted.
+
+The order of the last two terms are changed and two signs are swapped compared to Stange's
+paper to make the equivalence with elliptic relations unconditional (indepedent of W
+being an odd function). This should also avoid peculiarities in characteristic 3. -/
+def net (p q r s : ℤ) : R :=
+  W (p + q + s) * W (p - q) * W (r + s) * W r
+    - W (p + r + s) * W (p - r) * W (q + s) * W q
+    + W (q + r + s) * W (q - r) * W (p + s) * W p
+
+variable {W} in
+private lemma net_eq_rel₄ {p q r s : ℤ} :
+    net W p q r s = rel₄ W (2 * p + s) (2 * q + s) (2 * r + s) s := by
+  simp_rw [net, rel₄, addMulSub, add_add_add_comm _ s, add_sub_add_comm, sub_self, add_zero,
+    add_assoc, ← two_mul, add_sub_cancel_right, ← left_distrib, ← mul_sub_left_distrib,
+    Int.mul_div_cancel_left _ two_ne_zero]
+  ring
+
+/-- The three-index elliptic relation, obtained by
+specializing to `d = 0` in the four-index relation. -/
+def Rel₃ (m n r : ℤ) : Prop :=
+  W (m + n) * W (m - n) * W r ^ 2 =
+    W (m + r) * W (m - r) * W n ^ 2 - W (n + r) * W (n - r) * W m ^ 2
 
 /-- The proposition that a sequence indexed by integers is an elliptic sequence. -/
-def IsEllSequence : Prop :=
-  ∀ m n r : ℤ, W (m + n) * W (m - n) * W r ^ 2 =
-    W (m + r) * W (m - r) * W n ^ 2 - W (n + r) * W (n - r) * W m ^ 2
+def _root_.IsEllSequence : Prop :=
+  ∀ m n r : ℤ, Rel₃ W m n r
+
+/-- The numerator of an invariant of an elliptic sequence, such that for each `s`,
+`invarNum s n / invarDenom s n` is a constant independent of `n`. -/
+def invarNum (s n : ℤ) : R :=
+  (W (n + 2 * s) * W (n - s) ^ 2 + W (n + s) ^ 2 * W (n - 2 * s)) * W s ^ 2
+    + W n ^ 3 * W (2 * s) ^ 2
+
+/-- The denominator of an invariant of an elliptic sequence. -/
+def invarDenom (s n : ℤ) : R := W (n + s) * W n * W (n - s)
+
+theorem invarNum_mul_invarDenom_of_net (net_eq_zero : ∀ p q r s, net W p q r s = 0) (s m n : ℤ) :
+    invarNum W s m * invarDenom W s n = invarNum W s n * invarDenom W s m := by
+  linear_combination (norm := (simp_rw [invarNum, invarDenom, net]; ring_nf))
```

Comment:
My opinion on `ring_nf` is that it should never be used in the final code. Instead it's just a way to understand the expressions on both sides while proving a result, to be replaced with `ring` at the end (I prefer `ring1` because it would fail if both sides are not identical). It might be only my side but I also feel that `ring_nf` is significantly slower than `ring1` when compiling. Its behaviour should (heuristically) be identical to `ring1`, otherwise it's a bug...